src/main/java/org/opentripplanner/graph_builder/linking/SimpleStreetSplitter.java

```diff
@@ -96,30 +101,26 @@
 
     private static GeometryFactory geometryFactory = GeometryUtils.getGeometryFactory();
 
-    //If true edges are split and new edges are created (used when linking transit stops etc. during graph building)
-    //If false new temporary edges are created and no edges are deleted (Used when searching for origin/destination)
-    private final boolean destructiveSplitting;
-
     /**
      * Construct a new SimpleStreetSplitter.
      * NOTE: Only one SimpleStreetSplitter should be active on a graph at any given time.
      *
      * @param hashGridSpatialIndex If not null this index is used instead of creating new one
      * @param transitStopIndex Index of all transitStops which is generated in {@link org.opentripplanner.routing.impl.StreetVertexIndexServiceImpl}
-     * @param destructiveSplitting If true splitting is permanent (Used when linking transit stops etc.) when false Splitting is only for duration of a request. Since they are made from temporary vertices and edges.
      */
     public SimpleStreetSplitter(Graph graph, HashGridSpatialIndex<Edge> hashGridSpatialIndex,
-        SpatialIndex transitStopIndex, boolean destructiveSplitting) {
+        SpatialIndex transitStopIndex) {
         this.graph = graph;
         this.transitStopIndex = transitStopIndex;
-        this.destructiveSplitting = destructiveSplitting;
-        this.edgeFactory = new DefaultStreetEdgeFactory();
 
         //We build a spatial index if it isn't provided
         if (hashGridSpatialIndex == null) {
             // build a nice private spatial index, since we're adding and removing edges
             idx = new HashGridSpatialIndex<>();
             for (StreetEdge se : Iterables.filter(graph.getEdges(), StreetEdge.class)) {
+                // Note: Although it's up to the caller of idx.insert() to make it thread-safe, we don't need to
+                // worry about it here as it's inside the constructor of SimpleStreetSplitter
+                // and it's guaranteed that there is only one thread doing this.
```

Comment:
Please clarify how it is guaranteed that only one thread can call this constructor at once. It does look like it's only called in single-threaded contexts. The rules about graph updaters running sequentially in a single thread provide some guarantee there. But someone could easily call this constructor in a multi-threaded context. Maybe update the "note" in the Javadoc on this constructor to specifically say it must be called in a single thread only.

Reply:
Well, I didn't actually author this comment, so I have to agree with you here. I copied this verbatim from here: https://github.com/opentripplanner/OpenTripPlanner/pull/2596/files#diff-ff5b71da473283e2fbb823607004cf6c.

After looking this over rather thoroughly, it appears that even though there are 2 "notes" that say only 1 SimpleStreetSplitter should be in use per graph in the Javadoc of this class (1 in each constructor), this SimpleStreetSplitter class is being recreated numerous times on the same graph in various places, thus creating the potential for multiple inconsistent street indexes in the splitter. Given this issue, I am going to significantly revise this PR.

But even with these revisions, I still foresee that it would be easy for someone to perform thread-unsafe write modifications to a HashGridSpatialIndex instance despite there being notes in comments in various places about how you're not supposed to do that.

Reply:
It would be easy for someone to do it, but if they read the Javadoc on the classes and methods they are using, they should be warned not to do it. There are a lot of non-threadsafe data structures, and it's generally expected that someone will consult the documentation before mutating a data structure in multi-threaded code. That said, it doesn't hurt to add checks where feasible that contracts are being respected.

> this SimpleStreetSplitter class is being recreated numerous times on the same graph in various places, thus creating the potential for multiple inconsistent street indexes in the splitter.
I'm glad you spotted this fact, this clarifies somewhat what's going on here. It seems like this problematic code has been around for a long time in the bike parking updaters. Apparently the warning about having only one instance was not visible enough, and/or was not taken seriously because there was no explanation of why it was important to only have one. The warning should probably be moved from the SimpleStreetSplitter constructor to the class-level Javadoc, made more visible, and/or expanded with some explanation. And possibly duplicated on the constructor for good measure.

I think that class definitely needs some additional documentation also explaining what "destructive" vs. "non-destructive" splitting is, since it's not just about destroying edges, but also about whether they are indexed and therefore visible to subsequent searches and splitting operations.

Reply:
Ah, I see that destructive/non-destructive split documentation has been added in cf74d8132a2. Will review there.